baseplate/retry.py

```diff
@@ -0,0 +1,126 @@
+"""Policies for retrying an operation safely."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import time
+
+
+class RetryPolicy(object):
+    """A policy for retrying operations.
+
+    Policies are meant to be used as an iterable::
+
+        for time_remaining in RetryPolicy.new(attempts=3):
+            try:
+                some_operation.do(timeout=time_remaining)
+                break
+            except SomeError:
+                pass
+        else:
+            raise MaxRetriesError
+
+    """
+
+    def yield_attempts(self):
+        """Return an iterator which controls attempts.
+
+        On each iteration, the iterator will yield the number of seconds left
+        to retry, this should be used to set the timeout on the operation
+        being carried out. If there is no maximum time remaining,
+        :py:data:`None` is yielded instead.
+
+        The iterable will raise :py:exc:`StopIteration` once the operation
+        should not be retried any further.
+
+        """
+
+        raise NotImplementedError
+
+    def __iter__(self):
+        """Convenience alias for :py:meth:`yield_attempts`.
+
+        This allows policies to be directly iterated over.
+
+        """
+        return self.yield_attempts()
+
+    @staticmethod
+    def new(attempts=None, budget=None, backoff=None):
+        """Create a new retry policy with the given constraints.
+
+        :param int attempts: The maximum number of times the operation can be
+            attempted.
+        :param float budget: The maximum amount of time, in seconds, that the
+            local service will wait for the operation to succeed.
```

Comment:
> It might be worth making this "The maximum amount of time, in seconds, that the local service will wait for the operation to succeed, regardless of the number of attempts". This might be interpreted as budget per attempt, which is common API behavior (although in this case budget is used instead of the more common timeout parameter I've seen to indicate that). However, if I'm reading the code correctly, budget is a hard cap.

You're correct in your interpretation. Would `timeout` be a better name than `budget`?

> In the event that only attempts is specified, the timeout per attempt is managed at in the connection pool definition?

Yes. That's correct.